kyo-core/shared/src/main/scala/kyo/chunks.scala

```diff
@@ -0,0 +1,464 @@
+package kyo
+
+import Chunks.Indexed
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+sealed abstract class Chunk[T] derives CanEqual:
+
+    import Chunks.internal.*
+
+    //////////////////
+    // O(1) methods //
+    //////////////////
+
+    def size: Int
+
+    final def isEmpty: Boolean = size == 0
+
+    final def take(n: Int): Chunk[T] =
+        dropLeftAndRight(0, size - Math.min(Math.max(0, n), size))
+
+    final def dropLeft(n: Int): Chunk[T] =
+        dropLeftAndRight(Math.min(size, Math.max(0, n)), 0)
+
+    final def dropRight(n: Int): Chunk[T] =
+        dropLeftAndRight(0, Math.min(size, Math.max(0, n)))
+
+    final def slice(from: Int, until: Int): Chunk[T] =
+        dropLeftAndRight(Math.max(0, from), size - Math.min(size, until))
+
+    final def dropLeftAndRight(left: Int, right: Int): Chunk[T] =
+        @tailrec def loop(c: Chunk[T], left: Int, right: Int): Chunk[T] =
+            val size = c.size - left - right
+            if size <= 0 then Chunks.init
+            else
+                c match
+                    case Drop(chunk, dropLeft, dropRight, _) =>
+                        Drop(chunk, left + dropLeft, right + dropRight, size)
+                    case Append(chunk, value, size) if right > 0 =>
+                        loop(chunk, left, right - 1)
+                    case _ =>
+                        Drop(c, left, right, size)
+            end if
+        end loop
+        loop(this, left, right)
+    end dropLeftAndRight
+
+    final def append(v: T): Chunk[T] =
+        Append(this, v, size + 1)
```

Comment:
aside from adding complexity to run loops, is there a reason we cannot support `Append` with `Chunk[T]` member? It seems like a linked list of `Arrays` should be fairly efficient...

I understand there is some benefit in forcing amortization of re-allocating with the `concat` operator, but perhaps there is another way to do so?

Reply:
It is possible but if we add a `Concat` sub-class, all methods that do traversal need to use allocations to walk the structure since they'd have to handle a tail that is also a collection. With the current structure, all traversals are allocation-free even when the chunk isn't indexed. 

We could still attempt to include a `Concat` sub-class and use temporary buffers via thread locals to avoid allocations but there'd be some overhead as well. I think `concat` flattening the chunk to a contiguous array is a good property to avoid the data structure accumulating too many nested operations but I imagine we might revisit this decision if we see good evidence that `Concat` would be beneficial.